From 1c86adfd0be790983a4a616c4707fa20b8bb7fb1 Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Sat, 17 Nov 2018 08:34:35 +0100
Subject: [PATCH 1/1] fdt: sandbox: correct use of ${fdtcontroladdr}

The sandbox uses a virtual address space that is neither the physical nor
the virtual address space of the operating system. All address used on the
command line live in this address space. So also the environment variable
${fdtcontroladdr} has to be in this address space.

Commands like bootefi and booti receive the fdt address as parameter.
Without the patch ${fdtcontroladdr} cannot be used as parameter value on
the sandbox.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 common/board_r.c | 3 ++-
 lib/fdtdec.c     | 5 +++--
 2 files changed, 5 insertions(+), 3 deletions(-)

diff --git a/common/board_r.c b/common/board_r.c
index c55e33eec27..745212be71a 100644
--- a/common/board_r.c
+++ b/common/board_r.c
@@ -453,7 +453,8 @@ static int initr_env(void)
 	else
 		set_default_env(NULL, 0);
 #ifdef CONFIG_OF_CONTROL
-	env_set_addr("fdtcontroladdr", gd->fdt_blob);
+	env_set_hex("fdtcontroladdr",
+		    (unsigned long)map_to_sysmem(gd->fdt_blob));
 #endif
 
 	/* Initialize from environment */
diff --git a/lib/fdtdec.c b/lib/fdtdec.c
index a420ba18854..14b2220db1a 100644
--- a/lib/fdtdec.c
+++ b/lib/fdtdec.c
@@ -1252,8 +1252,9 @@ int fdtdec_setup(void)
 #  if CONFIG_IS_ENABLED(OF_PRIOR_STAGE)
 	gd->fdt_blob = (void *)prior_stage_fdt_address;
 #  else
-	gd->fdt_blob = (void *)env_get_ulong("fdtcontroladdr", 16,
-						(uintptr_t)gd->fdt_blob);
+	gd->fdt_blob = map_sysmem
+		(env_get_ulong("fdtcontroladdr", 16,
+			       (unsigned long)map_to_sysmem(gd->fdt_blob)), 0);
 #  endif
 # endif
 
-- 
2.19.1

